x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/openai/OpenAiResponseHandler.java

```diff
@@ -0,0 +1,131 @@
+import org.elasticsearch.xpack.inference.logging.ThrottlerManager;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+
+import static org.elasticsearch.core.Strings.format;
+import static org.elasticsearch.xpack.inference.external.http.HttpUtils.checkForEmptyBody;
+
+public class OpenAiResponseHandler implements ResponseHandler {
+
+    protected final String requestType;
+    private final CheckedFunction<HttpResult, List<? extends InferenceResults>, IOException> parseFunction;
+
+    public OpenAiResponseHandler(
+        String requestType,
+        CheckedFunction<HttpResult, List<? extends InferenceResults>, IOException> parseFunction
+    ) {
+        this.requestType = Objects.requireNonNull(requestType);
+        this.parseFunction = Objects.requireNonNull(parseFunction);
+    }
+
+    @Override
+    public void validateResponse(ThrottlerManager throttlerManager, Logger logger, HttpRequestBase request, HttpResult result)
+        throws RetryException {
+        checkForFailureStatusCode(request, result);
+        checkForEmptyBody(throttlerManager, logger, request, result);
+    }
+
+    @Override
+    public List<? extends InferenceResults> parseResult(HttpResult result) throws RetryException {
+        try {
+            return parseFunction.apply(result);
+        } catch (Exception e) {
+            throw new RetryException(true, e);
+        }
+    }
+
+    @Override
+    public String getRequestType() {
+        return requestType;
+    }
+
+    /**
+     * Validates the status code throws an RetryException if not in the range [200, 300).
+     *
+     * The OpenAI API error codes are document at https://platform.openai.com/docs/guides/error-codes/api-errors
+     * @param request The http request
+     * @param result  The http response and body
+     * @throws RetryException Throws if status code is {@code >= 300 or < 200 }
+     */
+    static void checkForFailureStatusCode(HttpRequestBase request, HttpResult result) throws RetryException {
+        int statusCode = result.response().getStatusLine().getStatusCode();
+        if (statusCode >= 200 && statusCode < 300) {
+            return;
+        }
+
+        // handle error codes
+        if (statusCode >= 500) {
+            String errorMsg = buildErrorMessageWithResponse(
+                "Received a server error status code for request [%s] status [%s]",
+                request.getRequestLine(),
+                statusCode,
+                result
+            );
+            throw new RetryException(false, errorMsg);
+        } else if (statusCode == 429) {
+            String errorMsg = buildErrorMessageWithResponse(
+                "Received a rate limit status code for request [%s] status [%s]",
+                request.getRequestLine(),
+                statusCode,
+                result
+            );
+            throw new RetryException(false, errorMsg); // TODO back off and retry
```

Comment:
Since the retrying sender already has exponential backoff should we return `true` here?